### PR TITLE
Enabling AMP on proper gfxArchs using DeviceProperties. 

### PR DESCRIPTION
**Pull request by stevenireeves**
Necessary changes for FI103.

### CODE REVIEW

File: tensorflow/core/grappler/devices.cc [hunk 1]
```diff
@@ -58,10 +59,13 @@ int GetNumAvailableGPUs(
...
             num_eligible_gpus++;
           }
-        }
-#else
-        num_eligible_gpus++;
+#else //TENSORFLOW_USE_ROCM
+          std::string gcn_arch_name = desc->rocm_amdgpu_gcn_arch_name(); 
+          if(std::find(std::begin(FP16SupportedDevices),
+             std::end(FP16SupportedDevices), gcn_arch_name)
+             != std::end(FP16SupportedDevices)) num_eligible_gpus++;
```
Comment (deven-amd) [inline tensorflow/core/grappler/devices.cc]: `GetNumAvailableGPUs` is a generic routine, and has nothing to do with whether or not the underlying GPU has fp16 support.  This routine should not require any changes...why was this change needed?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 2]
```diff
@@ -41,27 +41,35 @@ limitations under the License.
...
+bool GetFastFP16Support(const DeviceProperties &props)
+{
+    bool supported = false;
+    std::array<std::string, 2> FP16SupportedDevices = {"gfx906", "gfx908"};
+    std::string gcnArchName = props.environment().at("architecture");
+#if TF_ROCM_VERSION >= 4000
+    std::string gpu_arch = absl::StrSplit(gcnArchName, ":")[0];
+#else
+    std::string gpu_arch = gcnArchName;
+#endif
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: double check locally, but I think you can make the code within the `#if` unconditional.  `absl::StrSplit(gcnArchName, ":")[0]` should return you the gcnArchName, even when there is no ":" present

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 3]
```diff
@@ -41,27 +41,35 @@ limitations under the License.
...
-
-bool HasEnhancedFP16ComputeSupport(std::pair<int, int> gpu_arch){
-    std::string arch = std::to_string(gpu_arch.first); 
-    return std::find(std::begin(FP16SupportedDevices), 
-                     std::end(FP16SupportedDevices), arch)
-           != std::end(FP16SupportedDevices); 
+bool GetFastFP16Support(const DeviceProperties &props)
+{
+    bool supported = false;
+    std::array<std::string, 2> FP16SupportedDevices = {"gfx906", "gfx908"};
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: will a std::set<std:string> work better?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.h [hunk 4]
```diff
@@ -17,17 +17,17 @@ limitations under the License.
...
 namespace grappler {
 
-enum class AutoMixedPrecisionMode { CUDA, MKL };
-
-//Getting FP16 supported devices for ROCm
 #if TENSORFLOW_USE_ROCM
-bool HasEnhancedFP16ComputeSupport(std::pair<int, int> gpu_arch);
-#endif
+    bool GetFastFP16Support(const DeviceProperties& props); 
+#endif
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.h]: Put this after the enum decl for `AutoMixedPrecisionMode` to avoid the unnecessary diff....will be helpful when you file the upstream PR

Also make sure you have run clang-format

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc [hunk 5]
```diff
@@ -104,7 +102,7 @@ class AutoMixedPrecisionTest : public GrapplerTest {
     int num_gpus = GetNumAvailableGPUs();
     // If GPUs are available, require that they all satisfy the min arch.
     gpu_available_ = (num_gpus > 0);
-#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+#if GOOGLE_CUDA
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:105]: add a `#elif TENSORFLOW_USE_ROCM` here to unconditionally set `gpu_available_` to `false`.

That way you do not need to worry about modifying `GetNumAvailableGPUs` and this will still work as it will take the `else` block at line 111.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.h (hunk 4 above)
Comment (deven-amd) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.h]: actually we may not need this routine in the header file at all...see comments below

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc [hunk 6]
```diff
@@ -1095,8 +1074,8 @@ TEST_F(AutoMixedPrecisionTest, BatchMatMul) {
 #if GOOGLE_CUDA
   if (GetCudaVersion(*virtual_cluster_.get()) >= 9010) {
 #elif TENSORFLOW_USE_ROCM
-  if (HasEnhancedFP16ComputeSupport(GetMinDeviceGPUArch(*virtual_cluster_.get()))){
-#else  // TENSORFLOW_USE_ROCM
+  if (HasFP16Support(*virtual_cluster_.get())){
+#else
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc]: Once you make the change suggested above to have this test always run with `gfx906`, this check becomes moot, and you can revert to having `if (true)` as before.  
* It also means the `HasFP16Support` routine is not needed, which then implies
  * there will no more references to `GetFastFP16Support` in this file, which then implies
    * `GetFastFP16Support` can be kept local to the `auto_mixed_precision.cc` file!

File: tensorflow/core/util/rocm_solvers.cc [hunk 7]
```diff
@@ -111,7 +111,7 @@ struct ROCmSolverHandles {
   ~ROCmSolverHandles() {
     CHECK(wrap::rocblas_destroy_handle(parent_, rocm_blas_handle) ==
           rocblas_status_success)
-        << "Failed to destroy cuBlas instance.";
+        << "Failed to destroy rocBlas instance.";
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.cc]: good catch, but keep this in a separate commit. 

Keeping commits organized by functionality, helps "cherry-pick" them easier

**Review by deven-amd — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/devices.cc (hunk 1 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/devices.cc]: This was specifically used to get the number of qualified GPUs in the CUDA Implementation.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 2 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I believe you are correct.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 3 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I'm not sure it matters with the list this small. With std::array it's just a wrapper for a plain-old-data. Not sure if we want the overhead that comes with set for this size of list.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc (hunk 5 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:105]: Wouldn't we not  be actually testing builds with GPUs > gfx906 then?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: > I'm not sure it matters with the list this small

correct...the list is so small, that performance considerations are a no-op. Hence the argument for using std::set for better readability.

File: tensorflow/core/grappler/devices.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/core/grappler/devices.cc]: correct....the CUDA version gets the num availbale gpus that >= the specified cuda compute capability

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/grappler/devices.h#L29-L34

CUDA compute capability is a CUDA specific concept.

If you want something comparable for AMDGPUs, you would need something like

```
// Get the number of available AMDGPUs whose "gcn_arch" matches the given value
int GetNumAvailableAMDGPUs(const std::string& gcn_arch);
```

If you want something feature specific (like number of AMDGPUs that have FP16 support), you would need something like
```
// Get the number of available AMDGPUs with support for efficient FP16 compute
int GetNumAMDGPUsWithFP16Support();
```

Overloading this routine ( `GetNumAvailableGPUs`) to return the number of AMDGPUs which have efficient FP16 support is not right.
```

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc (hunk 5 above)
Comment (deven-amd) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:105]: This setup ensures that the AMP optimizer code gets enabled (without which all unit tests should fail).  That is the reason why we have code to create a dummy "gfx906" device, for the case where we do not have any GPUs that support efficient FP16 compute. Since that route will work even when we do have eligible GPUs, we can choose to take that route unconditionally...does that make sense?

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:105]: I understand.

**Comment by deven-amd** [comment]
m going to trigger the CI jobs manually, since we have finished moving the PR CI jobs in Jenkins

links the failing logs (before my manual trigger)
* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests-cpu/186/
* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests-gpu-multi/178/
* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests-gpu-single/175/
* http://ml-ci.amd.com:21096/job/tensorflow/job/github-prs-rocmfork-develop-upstream/job/rocm-latest-ubuntu-sanity/130/

**Comment by deven-amd** [comment]
jenkins : retest this please

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 8]
```diff
@@ -1990,10 +1994,16 @@ int GetNumGPUs(const Cluster& cluster,
...
+#if GOOGLE_CUDA
     std::pair<int, int> arch = GetDeviceGPUArch(device_properties);
     if (device_properties.type() == "GPU" && arch >= min_arch) {
       num_gpus++;
     }
+#elif TENSORFLOW_USE_ROCM
+    if (device_properties.type() == "GPU" && GetFastFP16Support(device_properties)){
+         num_gpus++; 
+    }
+#endif
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: @stevenireeves 
this logic is flawed in the case where `ShouldIgnorePerformance()` returns true
 (which will be the case for the `//tensorflow/python:auto_mixed_preceision_test_gpu` )

When `min_arch == {0, 0}` (i.e. `ShouldIgnorePerformance() == true`), 
* we need to return simply the number of GPUs, and 
* not the number of GPUs with efficient FP16 compute

Because kMinArch is set to `{0, 0}` for ROCm, you cannot use `min_arch == {0, 0}` check as a proxy for `ShouldIgnorePerformance() == true` (like it is done for the CUDA side).  You can get around this in one of two ways
1. Dont think we use kMinArch on the ROCm side at all, so you can let it have the same value as on the CUDA side, which will make it non-zero and the proxy check above, will work as expected
2. Explicitly pass in "ignore_performance" as a boolean flag to `GetNumGPUs`....this is overkill for CUDA, but makes things clear from a ROCm perspective

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: There was another solution. I could call ShouldIgnorePerformance directly in the TENSORFLOW_USE_ROCM conditional. That way I did not need to alter the args for GetNumGPUs. This solved the python: auto_mixed_precision failures locally.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 9]
```diff
@@ -41,27 +41,31 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/util/env_var.h"
 
+#if TENSORFLOW_USE_ROCM
+#include "rocm/include/hip/hip_runtime.h"
+#endif
+
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: do we need this include anymore?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 10]
```diff
@@ -1990,10 +1994,17 @@ int GetNumGPUs(const Cluster& cluster,
...
+#if GOOGLE_CUDA
     std::pair<int, int> arch = GetDeviceGPUArch(device_properties);
     if (device_properties.type() == "GPU" && arch >= min_arch) {
       num_gpus++;
     }
+#elif TENSORFLOW_USE_ROCM
+    if (ShouldIgnorePerformance()) num_gpus++; 
+    else if(device_properties.type() == "GPU" && GetFastFP16Support(device_properties)){
+         num_gpus++; 
+    }
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: still need to redo the logic.....num_gpus++ only if devcie == GPU

```
if (device_properties.type() == "GPU") {
  if (ShouldIgnorePerformance() || GetFastFP16Support(device_properties)) {
    num_gpus++;
  }
}
```

**Review by deven-amd — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 9 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I don't think we do.

**Review by deven-amd — APPROVED** [review]
